File: figures/src/python/classes/roc_pr_compound.py
#!/usr/bin/env python3
"""Class-based Python renderer for ROC and precision-recall compound figures."""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any

from matplotlib.figure import Figure

from figures.src.python.common import (
    REPO_ROOT,
    add_panel_labels,
    apply_publication_layout,
    configure_matplotlib,
    export_csv,
    mm_to_inches,
    project_resources,
    resolve_font_stack,
    source_data_mapping,
    validate_common_contract,
    write_manifest,
)


def _load_curve_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "model": row["model"],
                "panel": row["panel"],
                "x": float(row["x"]),
                "y": float(row["y"]),
                "y_lower": float(row["y_lower"]),
                "y_upper": float(row["y_upper"]),
                "operating_point": row["operating_point"],
            }
            for row in reader
        ]


def _load_metric_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        rows = [
            {
                "model": row["model"],
                "display_order": int(row["display_order"]),
                "auroc": float(row["auroc"]),
                "auprc": float(row["auprc"]),
                "ece": float(row["ece"]),
                "brier_score": float(row["brier_score"]),
                "prevalence": float(row["prevalence"]),
            }
            for row in reader
        ]
    return sorted(rows, key=lambda item: int(item["display_order"]))


def _source_rows(
    curves: list[dict[str, Any]], metrics: list[dict[str, Any]]
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    metric_map = {row["model"]: row for row in metrics}
    roc_rows: list[dict[str, Any]] = []
    pr_rows: list[dict[str, Any]] = []
    for row in curves:
        metric = metric_map[row["model"]]
        base = {
            "model": row["model"],
            "operating_point": row["operating_point"],
            "display_order": metric["display_order"],
            "ece": metric["ece"],
            "brier_score": metric["brier_score"],
        }
        if row["panel"] == "roc":
            roc_rows.append(
                {
                    **base,
                    "false_positive_rate": row["x"],
                    "true_positive_rate": row["y"],
                    "true_positive_rate_lower": row["y_lower"],
                    "true_positive_rate_upper": row["y_upper"],
                    "auroc": metric["auroc"],
                }
            )
        elif row["panel"] == "pr":
            pr_rows.append(
                {
                    **base,
                    "recall": row["x"],
                    "precision": row["y"],
                    "precision_lower": row["y_lower"],
                    "precision_upper": row["y_upper"],
                    "auprc": metric["auprc"],
                    "prevalence": metric["prevalence"],
                }
            )
    return roc_rows, pr_rows


def build_source_data(
    spec: dict[str, Any], curves: list[dict[str, Any]], metrics: list[dict[str, Any]]
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    roc_rows, pr_rows = _source_rows(curves, metrics)
    mapping = source_data_mapping(spec)
    export_csv(
        REPO_ROOT / mapping["a"],
        roc_rows,
        [
            "model",
            "display_order",
            "false_positive_rate",
            "true_positive_rate",
            "true_positive_rate_lower",
            "true_positive_rate_upper",
            "operating_point",
            "auroc",
            "ece",
            "brier_score",
        ],
    )
    export_csv(
        REPO_ROOT / mapping["b"],
        pr_rows,
        [
            "model",
            "display_order",
            "recall",
            "precision",
            "precision_lower",
            "precision_upper",
            "operating_point",
            "auprc",
            "prevalence",
            "ece",
            "brier_score",
        ],
    )
    return roc_rows, pr_rows


def _style_maps(theme: dict[str, Any], metrics: list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
    line_styles = ["solid", "--", "-."]
    markers = ["o", "s", "D"]
    colors = [
        theme["palette"]["categorical"][0],
        theme["palette"]["categorical"][1],
        theme["palette"]["categorical"][2],
    ]
    style_map: dict[str, dict[str, Any]] = {}
    for index, metric in enumerate(metrics):
        style_map[str(metric["model"])] = {
            "color": colors[index % len(colors)],
            "linestyle": line_styles[index % len(line_styles)],
            "marker": markers[index % len(markers)],
        }
    return style_map


def _label_offsets() -> dict[str, dict[str, tuple[float, float]]]:
    return {
        "roc": {
            "Foundation model": (0.03, 0.055),
            "Hybrid GNN": (0.03, -0.03),
            "CNN baseline": (0.03, -0.055),
        },
        "pr": {
            "Foundation model": (0.035, 0.045),
            "Hybrid GNN": (0.035, -0.015),
            "CNN baseline": (0.035, -0.05),
        },
    }


def _summary_text(metrics: list[dict[str, Any]], metric_key: str, title: str) -> str:
    lines = [title]
    for row in metrics:
        lines.append(f"{row['model']}: {row[metric_key]:.3f}")
    return "\n".join(lines)


def create_figure(spec_path: Path) -> Figure:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    curves = _load_curve_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    metrics = _load_metric_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    roc_rows, pr_rows = build_source_data(spec, curves, metrics)
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)

    import matplotlib.pyplot as plt

    fig, axes = plt.subplots(
        1,
        2,
        figsize=(
            mm_to_inches(float(spec["size"]["width_mm"])),
            mm_to_inches(float(spec["size"]["height_mm"])),
        ),
        constrained_layout=bool(theme["layout"]["constrained"]),
    )
    apply_publication_layout(fig, theme)
    style_map = _style_maps(theme, metrics)
    offsets = _label_offsets()

    panel_a = axes[0]
    for metric in metrics:
        model = str(metric["model"])
        model_rows = sorted(
            [row for row in roc_rows if str(row["model"]) == model],
            key=lambda item: float(item["false_positive_rate"]),
        )
        x_values = [float(row["false_positive_rate"]) for row in model_rows]
        y_values = [float(row["true_positive_rate"]) for row in model_rows]
        lower = [float(row["true_positive_rate_lower"]) for row in model_rows]
        upper = [float(row["true_positive_rate_upper"]) for row in model_rows]
        panel_a.fill_between(
            x_values,
            lower,
            upper,
            color=style_map[model]["color"],
            alpha=0.12,
            linewidth=0.0,
        )
        panel_a.plot(
            x_values,
            y_values,
            color=style_map[model]["color"],
            linestyle=style_map[model]["linestyle"],
            linewidth=1.8,
        )
        op_row = next(row for row in model_rows if str(row["operating_point"]) == "yes")
        panel_a.scatter(
            float(op_row["false_positive_rate"]),
            float(op_row["true_positive_rate"]),
            s=28,
            color=style_map[model]["color"],
            marker=style_map[model]["marker"],
            edgecolor=theme["palette"]["neutral"][0],
            linewidth=0.5,
            zorder=4,
        )
        dx, dy = offsets["roc"][model]
        panel_a.text(
            float(op_row["false_positive_rate"]) + dx,
            float(op_row["true_positive_rate"]) + dy,
            model,
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][0],
            ha="left",
            va="center",
        )
    panel_a.plot(
        [0.0, 1.0],
        [0.0, 1.0],
        color=theme["palette"]["neutral"][1],
        linestyle=":",
        linewidth=1.0,
    )
    panel_a.set_title("ROC discrimination", loc="left", pad=2)
    panel_a.set_xlabel("False positive rate")
    panel_a.set_ylabel("True positive rate")
    panel_a.set_xlim(0.0, 1.0)
    panel_a.set_ylim(0.0, 1.02)
    panel_a.spines["top"].set_visible(False)
    panel_a.spines["right"].set_visible(False)
    panel_a.text(
        0.98,
        0.05,
        _summary_text(metrics, "auroc", "AUROC summary"),
        transform=panel_a.transAxes,
        ha="right",
        va="bottom",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][0],
        bbox={
            "boxstyle": "round,pad=0.25",
            "facecolor": "white",
            "edgecolor": theme["palette"]["neutral"][2],
            "linewidth": 0.6,
        },
    )

    panel_b = axes[1]
    prevalence = float(metrics[0]["prevalence"])
    for metric in metrics:
        model = str(metric["model"])
        model_rows = sorted(
            [row for row in pr_rows if str(row["model"]) == model],
            key=lambda item: float(item["recall"]),
        )
        x_values = [float(row["recall"]) for row in model_rows]
        y_values = [float(row["precision"]) for row in model_rows]
        lower = [float(row["precision_lower"]) for row in model_rows]
        upper = [float(row["precision_upper"]) for row in model_rows]
        panel_b.fill_between(
            x_values,
            lower,
            upper,
            color=style_map[model]["color"],
            alpha=0.12,
            linewidth=0.0,
        )
        panel_b.plot(
            x_values,
            y_values,
            color=style_map[model]["color"],
            linestyle=style_map[model]["linestyle"],
            linewidth=1.8,
        )
        op_row = next(row for row in model_rows if str(row["operating_point"]) == "yes")
        panel_b.scatter(
            float(op_row["recall"]),
            float(op_row["precision"]),
            s=28,
            color=style_map[model]["color"],
            marker=style_map[model]["marker"],
            edgecolor=theme["palette"]["neutral"][0],
            linewidth=0.5,
            zorder=4,
        )
        dx, dy = offsets["pr"][model]
        panel_b.text(
            float(op_row["recall"]) + dx,
            float(op_row["precision"]) + dy,
            model,
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][0],
            ha="left",
            va="center",
        )
    panel_b.axhline(
        prevalence,
        color=theme["palette"]["neutral"][1],
        linestyle=":",
        linewidth=1.0,
    )
    panel_b.text(
        0.02,
        prevalence + 0.025,
        "Prevalence baseline",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
        ha="left",
        va="bottom",
    )
    panel_b.set_title("Precision-recall under imbalance", loc="left", pad=2)
    panel_b.set_xlabel("Recall")
    panel_b.set_ylabel("Precision")
    panel_b.set_xlim(0.0, 1.0)
    panel_b.set_ylim(0.0, 1.02)
    panel_b.spines["top"].set_visible(False)
    panel_b.spines["right"].set_visible(False)
    panel_b.text(
        0.98,
        0.05,
        _summary_text(metrics, "auprc", "AUPRC summary"),
        transform=panel_b.transAxes,
        ha="right",
        va="bottom",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][0],
        bbox={
            "boxstyle": "round,pad=0.25",
            "facecolor": "white",
            "edgecolor": theme["palette"]["neutral"][2],
            "linewidth": 0.6,
        },
    )

    add_panel_labels(spec, theme, [panel_a, panel_b])
    return fig


def build_figure(spec_path: Path) -> dict[str, str]:
    spec, _, font_policy, profile, _ = project_resources(spec_path)
    curves = _load_curve_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    metrics = _load_metric_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    build_source_data(spec, curves, metrics)
    resolved_font = resolve_font_stack(font_policy)
    fig = create_figure(spec_path)
    output_dir = REPO_ROOT / str(spec["renderers"]["python"]["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = str(spec["figure_id"])
    svg_path = output_dir / f"{stem}.svg"
    pdf_path = output_dir / f"{stem}.pdf"
    png_path = output_dir / f"{stem}.png"
    fig.savefig(svg_path, format="svg")
    fig.savefig(pdf_path, format="pdf")
    fig.savefig(png_path, format="png", dpi=int(profile["preview_dpi"]))
    outputs = {
        "svg": str(svg_path.relative_to(REPO_ROOT)),
        "pdf": str(pdf_path.relative_to(REPO_ROOT)),
        "png": str(png_path.relative_to(REPO_ROOT)),
    }
    write_manifest(
        spec,
        profile,
        resolved_font,
        "python",
        spec_path,
        outputs,
        [
            "paired_roc_and_pr_panels",
            "uncertainty_ribbons_for_curve_stability",
            "operating_point_markers_and_direct_labels",
            "prevalence_baseline_in_pr_panel",
            "vector_first_export",
        ],
        {
            "panel_count": len(spec.get("panels", [])),
            "reference_line_count": 2,
            "annotation_count": 8,
            "operating_point_count": 6,
            "uncertainty_band_count": 6,
        },
    )
    return outputs


if __name__ == "__main__":
    raise SystemExit("Use figures/src/python/run_class_renderer.py")

File: figures/src/r/classes/roc_pr_compound.R
repo_root <- getOption("manuscript.figure.repo_root")
if (is.null(repo_root)) {
  stop("manuscript.figure.repo_root option is required")
}
source(file.path(repo_root, "figures", "src", "r", "common.R"))

load_curve_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows$panel <- as.character(rows$panel)
  rows$model <- as.character(rows$model)
  rows$operating_point <- as.character(rows$operating_point)
  rows
}

load_metric_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows <- rows[order(rows$display_order), ]
  rows$model <- as.character(rows$model)
  rows
}

build_source_data <- function(spec, curves, metrics) {
  mapping <- source_data_mapping(spec)
  metric_map <- split(metrics, metrics$model)
  roc_rows <- data.frame()
  pr_rows <- data.frame()
  for (i in seq_len(nrow(curves))) {
    row <- curves[i, , drop = FALSE]
    metric <- metric_map[[row$model]][1, , drop = FALSE]
    base <- data.frame(
      model = row$model,
      display_order = metric$display_order,
      operating_point = row$operating_point,
      ece = metric$ece,
      brier_score = metric$brier_score,
      stringsAsFactors = FALSE
    )
    if (identical(row$panel, "roc")) {
      roc_rows <- rbind(
        roc_rows,
        cbind(
          base,
          false_positive_rate = row$x,
          true_positive_rate = row$y,
          true_positive_rate_lower = row$y_lower,
          true_positive_rate_upper = row$y_upper,
          auroc = metric$auroc,
          stringsAsFactors = FALSE
        )
      )
    } else if (identical(row$panel, "pr")) {
      pr_rows <- rbind(
        pr_rows,
        cbind(
          base,
          recall = row$x,
          precision = row$y,
          precision_lower = row$y_lower,
          precision_upper = row$y_upper,
          auprc = metric$auprc,
          prevalence = metric$prevalence,
          stringsAsFactors = FALSE
        )
      )
    }
  }
  write_source_csv(
    file.path(repo_root, mapping$a),
    roc_rows,
    c(
      "model",
      "display_order",
      "false_positive_rate",
      "true_positive_rate",
      "true_positive_rate_lower",
      "true_positive_rate_upper",
      "operating_point",
      "auroc",
      "ece",
      "brier_score"
    )
  )
  write_source_csv(
    file.path(repo_root, mapping$b),
    pr_rows,
    c(
      "model",
      "display_order",
      "recall",
      "precision",
      "precision_lower",
      "precision_upper",
      "operating_point",
      "auprc",
      "prevalence",
      "ece",
      "brier_score"
    )
  )
  list(roc = roc_rows, pr = pr_rows)
}

style_maps <- function(theme, metrics) {
  colors <- c(
    theme$palette$categorical[[1]],
    theme$palette$categorical[[2]],
    theme$palette$categorical[[3]]
  )
  line_types <- c("solid", "dashed", "dotdash")
  shapes <- c(16, 15, 18)
  data.frame(
    model = metrics$model,
    color = colors[seq_len(nrow(metrics))],
    line_type = line_types[seq_len(nrow(metrics))],
    shape = shapes[seq_len(nrow(metrics))],
    stringsAsFactors = FALSE
  )
}

label_offsets <- function() {
  data.frame(
    panel = c("roc", "roc", "roc", "pr", "pr", "pr"),
    model = c(
      "Foundation model",
      "Hybrid GNN",
      "CNN baseline",
      "Foundation model",
      "Hybrid GNN",
      "CNN baseline"
    ),
    dx = c(0.03, 0.03, 0.03, 0.035, 0.035, 0.035),
    dy = c(0.055, -0.03, -0.055, 0.045, -0.015, -0.05),
    stringsAsFactors = FALSE
  )
}

summary_text <- function(metrics, metric_column, title) {
  lines <- c(title)
  for (i in seq_len(nrow(metrics))) {
    lines <- c(lines, sprintf("%s: %.3f", metrics$model[[i]], metrics[[metric_column]][[i]]))
  }
  paste(lines, collapse = "\n")
}

create_plot <- function(spec_path) {
  spec <- load_yaml(spec_path)
  theme <- load_yaml(file.path(repo_root, "figures/config/project_theme.yml"))
  font_policy <- load_yaml(file.path(repo_root, "figures/config/font_policy.yml"))
  profiles <- load_yaml(file.path(repo_root, "figures/config/venue_profiles.yml"))
  registry <- load_yaml(file.path(repo_root, "figures/registry/classes.yml"))$classes
  profile <- profiles$profiles[[spec$target_profile]]
  class_entry <- registry[[spec$class_id]]
  validate_common_contract(spec, theme, font_policy, profile, class_entry)
  resolved_font <- resolve_font(font_policy)

  curves <- load_curve_rows(file.path(repo_root, spec$data_inputs[[1]]))
  metrics <- load_metric_rows(file.path(repo_root, spec$data_inputs[[2]]))
  source_rows <- build_source_data(spec, curves, metrics)
  styles <- style_maps(theme, metrics)
  offsets <- label_offsets()

  roc_rows <- merge(source_rows$roc, styles, by = "model", sort = FALSE)
  roc_labels <- merge(
    roc_rows[roc_rows$operating_point == "yes", ],
    offsets[offsets$panel == "roc", c("model", "dx", "dy")],
    by = "model",
    sort = FALSE
  )
  roc_labels$label_x <- roc_labels$false_positive_rate + roc_labels$dx
  roc_labels$label_y <- roc_labels$true_positive_rate + roc_labels$dy

  pr_rows <- merge(source_rows$pr, styles, by = "model", sort = FALSE)
  pr_labels <- merge(
    pr_rows[pr_rows$operating_point == "yes", ],
    offsets[offsets$panel == "pr", c("model", "dx", "dy")],
    by = "model",
    sort = FALSE
  )
  pr_labels$label_x <- pr_labels$recall + pr_labels$dx
  pr_labels$label_y <- pr_labels$precision + pr_labels$dy
  prevalence <- metrics$prevalence[[1]]

  plot_theme <- build_theme(theme, resolved_font) +
    ggplot2::theme(
      legend.position = "none",
      plot.margin = ggplot2::margin(5.5, 18, 5.5, 5.5)
    )

  plot_a <- ggplot2::ggplot(roc_rows, ggplot2::aes(x = false_positive_rate, y = true_positive_rate, color = model, linetype = model, group = model)) +
    ggplot2::geom_ribbon(
      ggplot2::aes(ymin = true_positive_rate_lower, ymax = true_positive_rate_upper, fill = model),
      alpha = 0.12,
      linewidth = 0,
      show.legend = FALSE
    ) +
    ggplot2::geom_line(linewidth = 0.8, show.legend = FALSE) +
    ggplot2::geom_abline(intercept = 0, slope = 1, linetype = "dotted", linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::geom_point(
      data = roc_rows[roc_rows$operating_point == "yes", ],
      ggplot2::aes(shape = model),
      size = 2.2,
      stroke = 0.35,
      fill = "white",
      show.legend = FALSE
    ) +
    ggplot2::geom_text(
      data = roc_labels,
      ggplot2::aes(x = label_x, y = label_y, label = model),
      inherit.aes = FALSE,
      hjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::annotate(
      "label",
      x = 0.98,
      y = 0.03,
      label = summary_text(metrics, "auroc", "AUROC summary"),
      hjust = 1,
      vjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      label.size = 0.25,
      fill = "white",
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::scale_color_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_fill_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_linetype_manual(values = stats::setNames(styles$line_type, styles$model)) +
    ggplot2::scale_shape_manual(values = stats::setNames(styles$shape, styles$model)) +
    ggplot2::labs(title = "ROC discrimination", x = "False positive rate", y = "True positive rate") +
    ggplot2::coord_cartesian(xlim = c(0, 1), ylim = c(0, 1.02), clip = "off") +
    plot_theme

  plot_b <- ggplot2::ggplot(pr_rows, ggplot2::aes(x = recall, y = precision, color = model, linetype = model, group = model)) +
    ggplot2::geom_ribbon(
      ggplot2::aes(ymin = precision_lower, ymax = precision_upper, fill = model),
      alpha = 0.12,
      linewidth = 0,
      show.legend = FALSE
    ) +
    ggplot2::geom_line(linewidth = 0.8, show.legend = FALSE) +
    ggplot2::geom_hline(yintercept = prevalence, linetype = "dotted", linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::annotate(
      "text",
      x = 0.02,
      y = prevalence + 0.025,
      label = "Prevalence baseline",
      hjust = 0,
      vjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::geom_point(
      data = pr_rows[pr_rows$operating_point == "yes", ],
      ggplot2::aes(shape = model),
      size = 2.2,
      stroke = 0.35,
      fill = "white",
      show.legend = FALSE
    ) +
    ggplot2::geom_text(
      data = pr_labels,
      ggplot2::aes(x = label_x, y = label_y, label = model),
      inherit.aes = FALSE,
      hjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::annotate(
      "label",
      x = 0.98,
      y = 0.03,
      label = summary_text(metrics, "auprc", "AUPRC summary"),
      hjust = 1,
      vjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      label.size = 0.25,
      fill = "white",
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::scale_color_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_fill_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_linetype_manual(values = stats::setNames(styles$line_type, styles$model)) +
    ggplot2::scale_shape_manual(values = stats::setNames(styles$shape, styles$model)) +
    ggplot2::labs(title = "Precision-recall under imbalance", x = "Recall", y = "Precision") +
    ggplot2::coord_cartesian(xlim = c(0, 1), ylim = c(0, 1.02), clip = "off") +
    plot_theme

  figure_plot <- (plot_a + plot_b + patchwork::plot_layout(ncol = 2)) +
    patchwork::plot_annotation(tag_levels = "a") &
    ggplot2::theme(
      plot.tag = ggplot2::element_text(
        size = as.numeric(theme$panel_labels$font_size_pt),
        face = theme$panel_labels$font_weight,
        family = resolved_font$family
      )
    )
  list(
    plot = figure_plot,
    spec = spec,
    theme = theme,
    profile = profile,
    resolved_font = resolved_font
  )
}

build_figure <- function(spec_path) {
  payload <- create_plot(spec_path)
  spec <- payload$spec
  theme <- payload$theme
  profile <- payload$profile
  resolved_font <- payload$resolved_font
  figure_plot <- payload$plot
  output_dir <- file.path(repo_root, spec$renderers$r$output_dir)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  stem <- spec$figure_id
  svg_path <- file.path(output_dir, sprintf("%s.svg", stem))
  pdf_path <- file.path(output_dir, sprintf("%s.pdf", stem))
  png_path <- file.path(output_dir, sprintf("%s.png", stem))
  width_in <- mm_to_inches(as.numeric(spec$size$width_mm))
  height_in <- mm_to_inches(as.numeric(spec$size$height_mm))

  svglite::svglite(file = svg_path, width = width_in, height = height_in, bg = theme$export_defaults$background, system_fonts = list(sans = resolved_font$family))
  print(figure_plot)
  grDevices::dev.off()
  grDevices::cairo_pdf(file = pdf_path, width = width_in, height = height_in, bg = theme$export_defaults$background, family = resolved_font$family)
  print(figure_plot)
  grDevices::dev.off()
  ragg::agg_png(filename = png_path, width = width_in, height = height_in, units = "in", res = as.numeric(profile$preview_dpi), background = theme$export_defaults$background)
  print(figure_plot)
  grDevices::dev.off()

  outputs <- list(
    svg = sub(paste0("^", repo_root, "/"), "", svg_path),
    pdf = sub(paste0("^", repo_root, "/"), "", pdf_path),
    png = sub(paste0("^", repo_root, "/"), "", png_path)
  )
  write_manifest(
    repo_root,
    spec,
    profile,
    resolved_font,
    spec_path,
    "r",
    outputs,
    list(
      "paired_roc_and_pr_panels",
      "uncertainty_ribbons_for_curve_stability",
      "operating_point_markers_and_direct_labels",
      "prevalence_baseline_in_pr_panel",
      "vector_first_export"
    ),
    list(
      panel_count = length(spec$panels),
      reference_line_count = 2,
      annotation_count = 8,
      operating_point_count = 6,
      uncertainty_band_count = 6
    )
  )
  outputs
}
